docs/serving/samples/grpc-ping-go/README.md

```diff
@@ -13,47 +13,101 @@ A simple gRPC server written in Go that you can use for testing.
-docker build \
-  --tag "docker.io/{username}/grpc-ping-go" \
-  --file=docs/serving/samples/grpc-ping-go/Dockerfile .
-docker push "docker.io/{username}/grpc-ping-go"
-```
+  ```shell
+  # Build the container on your local machine.
+  docker build --tag "{username}/grpc-ping-go" .
+
+  # Push the container to docker registry.
+  docker push "{username}/grpc-ping-go"
+  ```
+
+3. Update the `service.yaml` file in the project to reference the published image from step 1.
+
+   Replace `{username}` in `service.yaml` with your DockerHub username:
+   
+   
+  ```yaml
+  apiVersion: serving.knative.dev/v1alpha1
+  kind: Service
+  metadata:
+    name: grpc-ping
+    namespace: default
+  spec:
+    template:
+      spec:
+        containers:
+        - image: docker.io/{username}/grpc-ping-go
+          ports:
+            - name: h2c
+              containerPort: 8080
+  ``` 
+
+4. Use `kubectl` to deploy the service.
+
+  ```shell
+  kubectl apply --filename service.yaml
+  ```
 
-Next, replace `{username}` in `sample.yaml` with your DockerHub username, and
-apply the yaml.
+  Response:
+
+  ```shell
+  service "grpc-ping" created
+  ```
+
+## Exploring
+
+Once deployed, you can inspect the created resources with `kubectl` commands:
 
 ```shell
-kubectl apply --filename docs/serving/samples/grpc-ping-go/sample.yaml
+# This will show the Knative service that we created:
+kubectl get ksvc --output yaml
+
+# This will show the Route, created by the service:
+kubectl get route --output yaml
+
+# This will show the Configuration, created by the service:
+kubectl get configurations --output yaml
+
+# This will show the Revision, created by the Configuration:
+kubectl get revisions --output yaml
 ```
 
-## Use the client to stream messages to the gRPC server
+## Testing the service
+
+Testing the gRPC service uses the gRPC client built into the container.
 
 1. Fetch the created ingress hostname and IP.
 
    ```shell
    # Put the ingress IP into an environment variable.
-   export SERVICE_IP=`kubectl get svc istio-ingressgateway --namespace istio-system --output jsonpath="{.status.loadBalancer.ingress[*].ip}"`
+   export SERVICE_IP=$(kubectl get svc istio-ingressgateway --namespace istio-system --output jsonpath="{.status.loadBalancer.ingress[*].ip}")
    ```
 
-1. Use the client to send message streams to the gRPC server (replacing
-   `{username}`)
+1. Use the containerized client to send message streams to the gRPC server (replacing
+   `{username}`):
 
    ```shell
-   docker run -ti --entrypoint=/client docker.io/{username}/grpc-ping-go \
+   docker run --rm -it {username}/grpc-ping-go \
+     /client \
      -server_addr="${SERVICE_IP}:80" \
      -server_host_override="grpc-ping.default.example.com" \
      -insecure
    ```
+
+  The arguments after the container tag `{username}/grpc-ping-go` replace the container's default entrypoint which starts the server with a command that starts the gRPC client.
```

Comment:
This hard to understand.

Reply:
Which aspect is hard to understand (and it can be all of them):

* Swapping out the default entrypoint command
* Starting the gRPC client in the gRPC server container
* Something else?

Reply:
If "starts the server with a command that starts the gRPC client." == "Starting the gRPC client in the gRPC server container", then the new wording is better.

Reply:
I think this needs to be clearer, what do you think of this:

---

## Testing the service

Testing the gRPC service requires using a gRPC client built from the same protobuf definition
used by the server.

1. Fetch the created ingress hostname and IP.

   ```shell
   # Put the ingress IP into an environment variable.
   export SERVICE_IP=$(kubectl get svc istio-ingressgateway --namespace istio-system --output jsonpath="{.status.loadBalancer.ingress[*].ip}")
   ```

1. Use the gRPC client to send message streams to the gRPC server.

   The Dockerfile builds the client binary. To run the client you will use the
   same container image deployed for the server with an override to the
   entrypoint command to use the client binary instead of the server binary.

   Replace `{username}` with your Docker Hub user name and run the command:

   ```shell
   docker run --rm -it {username}/grpc-ping-go \
     /client \
     -server_addr="${SERVICE_IP}:80" \
     -server_host_override="grpc-ping.default.example.com" \
     -insecure
   ```

  The arguments after the container tag `{username}/grpc-ping-go` are used
  instead of the entrypoint command defined in the Dockerfile `CMD` statement.

Reply:
That is great!